pkg/tfbridge/walk.go

```diff
@@ -168,3 +265,347 @@ func LookupSchemaInfoPath(schemaPath SchemaPath, schemaInfo *SchemaInfo) *Schema
+func (r VisitDataSourceRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitDataSourceRoot) isVisitRoot()             {}
+
+type VisitConfigRoot struct{ token tokens.Type }
+
+func (r VisitConfigRoot) PulumiToken() tokens.Type { return r.token }
+func (r VisitConfigRoot) isVisitRoot()             {}
+
+type Visitor = func(VisitInfo, SchemaPath) (VisitResult, error)
+
+func MustTraverseSchemas(prov *ProviderInfo, traversalID string, visitor Visitor) {
+	err := TraverseSchemas(prov, traversalID, visitor)
+	contract.AssertNoErrorf(err, "failed to traverse provider schemas")
+}
+
+// Traverse all schemas in a provider.
+//
+// traversalId must be unique for each traversal within a `make tfgen`, but the same
+// between `make tfgen` and the provider runtime. A simple descriptive string like `"apply
+// labels"` works great here.
+func TraverseSchemas(prov *ProviderInfo, traversalID string, visitor Visitor) error {
+	prov.MetadataInfo.assertValid() // We must have metadata info for an efficient (sparse) traversal.
+
+	var errs []error
+
+	const key = "traverse-effects"
+	forEffect := currentRuntimeStage == runningProviderStage
+	var effects traversalRecord
+
+	// We are in forEffect mode, so we need to load in the list of paths where we need
+	// to replay effects.
+	if forEffect {
+		var ok bool
+		var err error
+		effects, ok, err = md.Get[traversalRecord](prov.GetMetadata(), key)
+		contract.Assertf(ok, "could not find key %s in metadata, are you missing a `make tfgen`?", key)
+		if err != nil {
+			return fmt.Errorf("unable to find effect list: %w", err)
+		}
+	}
+
+	rMap := prov.P.ResourcesMap()
+	for tfName, rInfo := range prov.Resources {
+		rShim, ok := rMap.GetOk(tfName)
+		if !ok {
+			continue
+		}
+
+		var err error
+
+		effect := ensureMap(&effects.Resources)[tfName]
+
+		effect, err = traverseResourceOrDataSource(tfName, rShim, rInfo,
+			visitor, forEffect, effect)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		effects.Resources[tfName] = effect
+	}
+
+	dMap := prov.P.DataSourcesMap()
+	for tfName, dInfo := range prov.DataSources {
+		rShim, ok := dMap.GetOk(tfName)
+		if !ok {
+			continue
+		}
+
+		var err error
+		effect := ensureMap(&effects.DataSources)[tfName]
+
+		effect, err = traverseResourceOrDataSource(tfName, rShim, dInfo,
+			visitor, forEffect, effect)
+		if err != nil {
+			errs = append(errs, err)
+		}
+		effects.DataSources[tfName] = effect
+	}
+
+	{
+		var err error
+		effects.Config, err = traverseSchemaMap(prov.P.Schema(), ensureMap(&prov.Config),
+			VisitConfigRoot{tokens.Type("pulumi:providers" + prov.Name)}, visitor, forEffect, effects.Config)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	// We are in !forEffect mode, so we need to save the list of effects to generate.
+	if !forEffect {
+		effects.clean()
+		err := md.Set(prov.GetMetadata(), key, effects)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errors.Join(errs...)
+}
+
+type traversalRecord struct {
```

Comment:
There is some non-trivial TCO to keeping things serialized. Why is this code storing this? What does this data mean?

Reply:
I'll add a doc comment explaining. The tldr is that `traversalRecord` is a list of resource/datasource/config schema paths that need to be visited during runtime.

I added this text to the PR description:
> The implementation takes after https://github.com/pulumi/pulumi-terraform-bridge/pull/1544. Visitors return if they had an "effect". During make tfgen, every path is visited and a list of "effectful" SchemaPaths is built. During the provider's normal runtime, this list is deserialized and only these paths are visited.